src/osd/osd_types.cc

```diff
@@ -1960,6 +1987,47 @@ void pg_stat_t::generate_test_instances(list<pg_stat_t*>& o)
   o.push_back(new pg_stat_t(a));
 }
 
+bool operator==(const pg_stat_t& l, const pg_stat_t& r)
+{
+  return
+    l.version == r.version &&
+    l.reported_seq == r.reported_seq &&
+    l.reported_epoch == r.reported_epoch &&
+    l.state == r.state &&
+    l.last_fresh == r.last_fresh &&
+    l.last_change == r.last_change &&
+    l.last_active == r.last_active &&
+    l.last_clean == r.last_clean &&
+    l.last_unstale == r.last_unstale &&
+    l.last_undegraded == r.last_undegraded &&
+    l.last_fullsized == r.last_fullsized &&
+    l.log_start == r.log_start &&
+    l.ondisk_log_start == r.ondisk_log_start &&
+    l.created == r.created &&
+    l.last_epoch_clean == r.last_epoch_clean &&
+    l.parent == r.parent &&
+    l.parent_split_bits == r.parent_split_bits &&
+    l.last_scrub == r.last_scrub &&
+    l.last_deep_scrub == r.last_deep_scrub &&
+    l.last_scrub_stamp == r.last_scrub_stamp &&
+    l.last_deep_scrub_stamp == r.last_deep_scrub_stamp &&
+    l.last_clean_scrub_stamp == r.last_clean_scrub_stamp &&
+    l.stats == r.stats &&
+    l.stats_invalid == r.stats_invalid &&
+    l.log_size == r.log_size &&
+    l.ondisk_log_size == r.ondisk_log_size &&
+    l.up == r.up &&
+    l.acting == r.acting &&
+    l.mapping_epoch == r.mapping_epoch &&
+    l.blocked_by == r.blocked_by &&
+    l.last_became_active == r.last_became_active &&
+    l.dirty_stats_invalid == r.dirty_stats_invalid &&
+    l.omap_stats_invalid == r.omap_stats_invalid &&
+    l.hitset_stats_invalid == r.hitset_stats_invalid &&
+    l.hitset_bytes_stats_invalid == r.hitset_bytes_stats_invalid &&
+    l.up_primary == r.up_primary &&
+    l.acting_primary == r.acting_primary;
+}
```

Comment:
On Wed, 21 Jan 2015, Haomai Wang wrote:
> 
> In src/osd/osd_types.cc:
> 
> > +    l.stats == r.stats &&
> > +    l.stats_invalid == r.stats_invalid &&
> > +    l.log_size == r.log_size &&
> > +    l.ondisk_log_size == r.ondisk_log_size &&
> > +    l.up == r.up &&
> > +    l.acting == r.acting &&
> > +    l.mapping_epoch == r.mapping_epoch &&
> > +    l.blocked_by == r.blocked_by &&
> > +    l.last_became_active == r.last_became_active &&
> > +    l.dirty_stats_invalid == r.dirty_stats_invalid &&
> > +    l.omap_stats_invalid == r.omap_stats_invalid &&
> > +    l.hitset_stats_invalid == r.hitset_stats_invalid &&
> > +    l.hitset_bytes_stats_invalid == r.hitset_bytes_stats_invalid &&
> > +    l.up_primary == r.up_primary &&
> > +    l.acting_primary == r.acting_primary;
> > +}
> 
> Why not use memcmp(&l, &r, sizeof(pg_stat_t))? Maybe more faster and safe for later changes?

Then we need to make the structure packed (or be very careful about 
alignment).  But yeah that is probably less prone to breakage!